aspnetcore/blazor/security/server/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,143 @@
+
+In the `App` component (`Components/App.razor`), place a library script reference after [Blazor's `<script>` tag](xref:blazor/project-structure#location-of-the-blazor-script):
+
+```razor
+<script src="qrcode.min.js"></script>
+```
+
+The `EnableAuthenticator` component, which is part of the QR code system in the app and displays the QR code to users, adopts static server-side rendering (static SSR) with enhanced navigation. Therefore, normal scripts can't execute when the component loads or updates under enhanced navigation. Extra steps are required to trigger the QR code to load in the UI when the page is loaded. To accomplish loading the QR code, the approach explained in <xref:blazor/js-interop/ssr> is adopted.
+
+Add the following [JavaScript initializer](xref:blazor/fundamentals/startup#javascript-initializers) to the server project's `wwwroot` folder. The `{NAME}` placeholder must be the name of the app's assembly in order for Blazor to locate and load the file automatically. If the server app's assembly name is `BlazorSample`, the file is named `BlazorSample.lib.module.js`.
+
+`wwwroot/{NAME}.lib.module.js`:
+
+[!INCLUDE[](~/blazor/includes/js-interop/blazor-page-script.md)]
+
+Add the following shared `PageScript` component to the server app.
+
+`Components/PageScript.razor`:
+
+```razor
+<page-script src="@Src"></page-script>
+
+@code {
+    [Parameter]
+    [EditorRequired]
+    public string Src { get; set; } = default!;
+}
+```
+
+Add the following [collocated JS file](xref:blazor/js-interop/index#load-a-script-from-an-external-javascript-file-js-collocated-with-a-component) for the `EnableAuthenticator` component, which is located at `Components/Account/Pages/Manage/EnableAuthenticator.razor`. The `onLoad` function creates the QR code with Sangmin's `qrcode.js` library using the QR code URI produced by the `GenerateQrCodeUri` method in the component's `@code` block.
+
+`Components/Account/Pages/Manage/EnableAuthenticator.razor.js`:
+
+```javascript
+export function onLoad() {
+  const uri = document.getElementById('qrCodeData').getAttribute('data-url');
+  new QRCode(document.getElementById('qrCode'), uri);
+}
+
+export function onUpdate() { }
+
+export function onDispose() { }
+```
+
+Under the `<PageTitle>` component in the `EnableAuthenticator` component, add the `PageScript` component with the path to the collocated JS file:
+
+```razor
+<PageScript Src="./Components/Account/Pages/Manage/EnableAuthenticator.razor.js" />
+```
+
+> [!NOTE]
+> An alternative to using the approach with the `PageScript` component is to use an event listener (`Blazor.addEventListener("enhancedload", {CALLBACK})`) registered in an [`afterWebStarted` JS initializer](xref:blazor/fundamentals/startup#javascript-initializers) to listen for page updates caused by enhanced navigation. The callback (`{CALLBACK}` placeholder) performs the QR code initialization logic.
```

Comment:
```suggestion
> An alternative to using the approach with the `PageScript` component is to use an event listener (`blazor.addEventListener("enhancedload", {CALLBACK})`) registered in an [`afterWebStarted` JS initializer](xref:blazor/fundamentals/startup#javascript-initializers) to listen for page updates caused by enhanced navigation. The callback (`{CALLBACK}` placeholder) performs the QR code initialization logic.
```

(because if this is happening in the `afterWebStarted` callback, then `blazor` is passed as an argument, so it would be preferable to use the argument instead of the `window.Blazor` object)